### PR TITLE
[NCLSUP-990] -  Do not create duplicated artifacts after DelAnalyzer analysis

### PR DESCRIPTION
**Pull request by vibe13**
### Checklist:

* [ ] Have you added unit tests for your change?

### CODE REVIEW

File: facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java [hunk 1]
```diff
@@ -242,22 +249,85 @@ public void clear(int id) {
...
-    private org.jboss.pnc.model.Artifact findOrCreateNotFoundArtifact(Artifact artifact, TargetRepository targetRepo) {
-        return findOrCreateArtifact(mapNotFoundArtifact(artifact), targetRepo);
+    private Optional<org.jboss.pnc.model.Artifact> getBestMatchingArtifact(
+            Collection<org.jboss.pnc.model.Artifact> artifacts) {
+        if (artifacts == null || artifacts.isEmpty()) {
+            return Optional.empty();
+        }
+
+        if (artifacts.size() == 1) {
+            return Optional.of(artifacts.iterator().next());
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java:259]: If you want it shorter :smile: 
```suggestion
            return artifacts.stream().findFirst();
```

File: facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java [hunk 2]
```diff
@@ -242,22 +249,85 @@ public void clear(int id) {
...
+        }
+
+        return Optional.of(
+                artifacts.stream()
+                        .sorted(Comparator.comparing(DeliverableAnalyzerManagerImpl::getArtifactQuality).reversed())
+                        .findFirst()
+                        .orElse(artifacts.iterator().next()));
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java:266]: What is this supposed to do? `artifacts` is not empty, so findFirst should always be present Optional.

File: facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java [hunk 3]
```diff
@@ -242,22 +249,85 @@ public void clear(int id) {
...
     }
 
-    private org.jboss.pnc.model.Artifact findOrCreateArtifact(
+    private static int getArtifactQuality(Object obj) {
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java:269]:
```suggestion
    private static int getNotFoundArtifactRating(org.jboss.pnc.model.Artifact artifact) {
```

**Review by janinko — CHANGES_REQUESTED** [review]

**Comment by vibe13** [comment]
Thanks for the review and improvements @janinko ! I will wait for https://github.com/project-ncl/pnc/pull/3875 to land first. In the meantime I will make the same changes in https://github.com/project-ncl/pnc/pull/3864. Thanks!

**Comment by vibe13** [comment]
retest this please

File: facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java [hunk 4]
```diff
@@ -316,22 +324,83 @@ public void clear(int id) {
...
+            return Optional.empty();
+        }
+
+        if (artifacts.size() == 1) {
+            return artifacts.stream().findFirst();
+        }
+
+        return artifacts.stream()
+                .sorted(Comparator.comparing(DeliverableAnalyzerManagerImpl::getNotFoundArtifactRating).reversed())
+                .findFirst();
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java:339]: Looking at this, the special case for size 1 is not technically necessary. But maybe it's more understandable like this?

File: facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java [hunk 5]
```diff
@@ -316,22 +324,83 @@ public void clear(int id) {
...
     }
 
-    private org.jboss.pnc.model.Artifact findOrCreateArtifact(
+    private static int getNotFoundArtifactRating(Object obj) {
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java:342]: Why the method can't take Artifact?

**Pull request closed without merging**